jps/model-impl/src/com/intellij/openapi/fileTypes/impl/FileTypeAssocTable.java

```diff
@@ -84,14 +93,24 @@ public T addAssociation(@NotNull FileNameMatcher matcher, @NotNull T type) {
       Map<CharSequence, T> mapToUse = exactFileNameMatcher.isIgnoreCase() ? myExactFileNameAnyCaseMappings : myExactFileNameMappings;
       return mapToUse.put(exactFileNameMatcher.getFileName(), type);
     }
-    int i = ContainerUtil.indexOf(myMatchingMappings, p -> p.first.equals(matcher));
-    if (i == -1) {
-      myMatchingMappings.add(Pair.create(matcher, type));
-      return null;
+
+    // Since we need to sort on insertion for IJPL-149806, we allow only one thread to modify the mapping list at a time
+    myMatchingMappingsLock.lock();
```

Comment:
please get rid of this lock, it's unnecessary

Reply:
Is it? Granted, I made a wrong assumption about stdlib's COW arraylist, it seems that it also overrides `removeIf` to be atomic (c.f. https://github.com/openjdk/jdk/blob/jdk-17%2B35/src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java#L814-L869), so at least from the sorting perspective it seems to be unnecessary — my bad.

But what if some other thread modifies the list between `indexOf` and `add` — I see there's `addIfAbsent` that's atomic, but it would need a predicate version, because we want to compare by `first.equals` instead of just `equals` — won't we duplicate items then?

Or what if some other thread modifies the list between `indexOf` and `set`? Won't we overwrite a wrong matching then?

Say, one thread runs `addAssociation` while another runs `removeAssociation` and it clobbers the list for `addAssociation`?

If you're sure that can _never_ happen, then the lock if probably not needed, but I don't have enough idea about what can use this class and if modifications can happen concurrently from different threads, so I thought it's better to be defensive. If you're sure that's not necessary, then I guess it can be removed. I just want to be sure I'm not missing something here.

Reply:
myMatchingMappings is changed only during showing of the modal dialog in Settings|file types. there's no need to guard the modification any further, for simplicity/readability reasons.